File: 프로그래머스/LV2/코딩테스트 연습/깊이,너비 우선 탐색(DFS,BFS)/타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int go(vector<int>& numbers, int target, int i, int s){
    int ans = 0;
    if(i == numbers.size())return s == target ? 1: 0;

    ans += go(numbers, target, i+1, s+numbers[i]);
    ans += go(numbers, target, i+1, s-numbers[i]);
    return ans;
}

int solution(vector<int> numbers, int target){
    return go(numbers, target, 0, 0);
}

File: 프로그래머스/LV2/코딩테스트 연습/깊이,너비 우선 탐색(DFS,BFS)/타겟 넘버/타겟 넘버.py
def go(numbers, target, i, s):
    ans = 0

    if i == len(numbers):
        if s == target:
            return 1
        return 0
    
    ans += go(numbers, target, i+1, s+numbers[i])
    ans += go(numbers, target, i+1, s-numbers[i])
    return ans

def solution(numbers, target):    
    return go(numbers, target, 0, 0)
